File: problems/linked_list_intersection/linked_list_intersection.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def getIntersectionNode(
        self, headA: ListNode, headB: ListNode
    ) -> Optional[ListNode]:
        one = headA
        two = headB

        while one != two:
            one = one.next if one else headB
            two = two.next if two else headA
        return one

File: problems/linked_list_intersection/LinkedListInterSection.java
package problems.linked_list_intersection;

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }
}

public class LinkedListInterSection {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode one = headA;
        ListNode two = headB;

        while (one != two) {
            one = (one != null) ? one.next : headB;
            two = (two != null) ? two.next : headA;
        }

        return one;
    }
}
